VIP 아키텍처를 적용해 홈 화면 구성

song1900 opened the pull request:
VIP 아키텍처를 적용해 HomeViewController와 관련 컴포넌트를 구현했습니다.
( 현재는 단순 뷰 전환 기능만 있어 데이터는 존재 하지 않습니다.)

- HomeViewController - 인터페이스 로직을 처리하고, Interactor에서 데이터를 받아서 화면에 표시합니다.
- HomeInteractor - 비즈니스 로직을 수행합니다.
- HomePresenter - Interactor로부터 받은 데이터를 View에 전달합니다. 

HomeViewController는 화면 전체를 CollectionView로 구성되어 있고 첫 번째 셀에 HomeSearchView가 존재합니다.
검색 바를 클릭하면 새로운 화면으로 전환됩니다.

song1900 committed:
Refactor: Home 아키텍처 변경

MVVM 아키텍처로 구성된 Home을 VIP 아키텍처 구조로 변경

song1900 committed:
Chore: 필요 없는 파일 제거

song1900 committed:
Feat: BaseViewController 생성

공통 기능인 키보드 숨기기를 포함하는 뷰 컨트롤러를 생성했습니다.
해당 기능이 필요한 뷰 컨트롤러에서 이 클래스를 상속받도록 함으로써, 중복 코드 없이 기능을 제공하도록 했습니다.

song1900 committed:
Feat: NavigationBarHideable 프로토콜 추가

네비게이션 바 숨기고 보여주기 기능을 위한 프로토콜

song1900 committed:
Chore: 필요없는 파일 제거 및 constraint 수정

song1900 committed:
Feat: Home 화면 Search Cell, SearchBar View 생성

song1900 committed:
Feat: Home화면 VIP 아키텍처를 적용해 화면 구현

song1900 committed:
Chore: 홈 화면 다국어 지원 추가

song1900 committed:
Merge branch 'feature/signup-registration' of https://github.com/f-lab-edu/gil-navi-ios into feature/setup-home-screen

# Conflicts:
#	GIL/GIL.xcodeproj/project.pbxproj

f-lab-jercy commented on GIL/GIL/Screen/Home/HomeViewController.swift:
```diff
@@ -30,25 +36,114 @@ final class HomeViewController: UIViewController {
+        5
+    }
+    
+    func collectionView(
+        _ collectionView: UICollectionView,
+        cellForItemAt indexPath: IndexPath
+    ) -> UICollectionViewCell {
+        switch indexPath.row {
+        case 0:
+            if let cell = collectionView.dequeueReusableCell(withReuseIdentifier: HomeSearchCollectionViewCell.reuseIdentifier, for: indexPath) as? HomeSearchCollectionViewCell {
+                
+                cell.onSearchBarTapped = { [weak self] in
+                    guard let self else { return }
+                    self.interactor.performSearch()
+                }
+                
+                return cell
+            }
+            
+        default: break
+        }
+        
+        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "NormalCell", for: indexPath)
+        cell.backgroundColor = .gray
+        
+        return cell
```
compositionalLayout, diffableDatasource로 추가 적용을 해보면 좋을거 같습니다.

song1900 replied:
https://github.com/f-lab-edu/gil-navi-ios/pull/17/commits/4fe62ed7f6f70fb45103f79ad6ec61a0ed829a88
해당 커밋으로 수정했습니다!

f-lab-jercy commented on GIL/GIL/Screen/Home/HomeInteractor.swift:
```diff
@@ -0,0 +1,20 @@
+final class HomeInteractor: HomeBusinessLogic {
+    var presenter: HomePresentationLogic?
+    
+    func performSearch() {
+        presenter?.presentSearchScreen()
+    }
```
역햘들이 조금 더 추가 되었을때 한번더 확인해보겠습니다.

song1900 committed:
Merge branch 'main' of https://github.com/f-lab-edu/gil-navi-ios into feature/setup-home-screen

song1900 committed:
Refactor: HomeView CollectionView 리팩토링

- "UICollectionViewCompositionalLayout", "DiffableDataSource"를 사용해 컬렉션 뷰 구성
- 10-200 룰 적용
